tracing-flame/src/lib.rs

```diff
@@ -0,0 +1,273 @@
+//!
+//! ```console
+//! # flamegraph
+//! cat tracing.folded | inferno-flamegraph > tracing-flamegraph.svg
+//!
+//! #flamechart
+//! cat tracing.folded | inferno-flamegraph --flamechart > tracing-flamechart.svg
+//! ```
+//!
+//! ## Differences between `flamegraph`s and `flamechart`s
+//!
+//! By default, `inferno-flamegraph` creates flamegraphs. Flamegraphs operate by
+//! that collapsing identical stack frames and sorting them on the frame's names.
+//! 
+//! This behavior is great for multithreaded programs and long-running programs
+//! where the same frames occur _many_ times, for short durations, because it reduces
+//! noise in the graph and gives the reader a better idea of the 
+//! overall time spent in each part of the application.
+//!
+//! However, it is sometimes desirable to preserve the _exact_ ordering of events
+//! as they were emitted by `tracing-flame`, so that it is clear when each
+//! span is entered relative to others and get an accurate visual trace of
+//! the execution of your program. This representation is best created with a 
+//! `flamechart`, which _does not_ sort or collapse identical stack frames.
+//!
+//! [`tracing`]: https://docs.rs/tracing
+//! [`inferno`]: https://docs.rs/inferno
+//! [`FlameLayer`]: struct.FlameLayer.html
+mod error;
+
+use error::Error;
+use std::fmt;
+use std::fmt::Write as _;
+use std::fs::File;
+use std::io::BufWriter;
+use std::io::Write;
+use std::marker::PhantomData;
+use std::path::Path;
+use std::sync::Mutex;
+use std::time::{Duration, Instant};
+use tracing::span::Attributes;
+use tracing::Id;
+use tracing::Subscriber;
+use tracing_subscriber::layer::Context;
+use tracing_subscriber::registry::LookupSpan;
+use tracing_subscriber::registry::SpanRef;
+use tracing_subscriber::Layer;
+use std::sync::Arc;
+
+pub struct FlameLayer<S, W>
+where
+    S: Subscriber,
+    W: Write + 'static,
+{
+    out: Arc<Mutex<W>>,
+    last_event: Mutex<Instant>,
+    _inner: PhantomData<S>,
+}
+
+pub struct FlameGuard<W>
+where
+    W: Write + 'static,
+{
+    out: Arc<Mutex<W>>,
+}
+
+impl<S, W> FlameLayer<S, W>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+    W: Write + 'static,
+{
+    pub fn new(writer: W) -> Self {
+        Self {
+            out: Arc::new(Mutex::new(writer)),
+            last_event: Mutex::new(Instant::now()),
+            _inner: PhantomData,
+        }
+    }
+
+    pub fn flush_on_drop(&self) -> FlameGuard<W> {
+        FlameGuard {
+            out: self.out.clone(),
+        }
+    }
+}
+
+impl<W> Drop for FlameGuard<W>
+where
+    W: Write + 'static,
+{
+    fn drop(&mut self) {
+        dbg!(());
+        match self.out.lock().unwrap().flush() {
+            Ok(_) => (),
+            Err(e) => eprintln!("Error: {}", e),
+        }
+    }
+}
+
+impl<S> FlameLayer<S, BufWriter<File>>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+{
+    pub fn write_to_file(path: impl AsRef<Path>) -> Result<Self, Error> {
```

Comment:
nit/tioli: this name doesn't really make it clear to me that this is a constructor...how about 
```suggestion
    pub fn with_file(path: impl AsRef<Path>) -> Result<Self, Error> {
```
```suggestion
    pub fn with_output_file(path: impl AsRef<Path>) -> Result<Self, Error> {
```
or something?